From 28de3d04af07a77673295266b246caef174b7eeb Mon Sep 17 00:00:00 2001
From: oren <oren@blueskyscientific.com>
Date: Tue, 1 Aug 2017 18:27:51 -0700
Subject: [PATCH] testing patch

---
 include/sound/tlv320aic3x.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/sound/tlv320aic3x.h b/include/sound/tlv320aic3x.h
index 9407fd0..421ad03 100644
--- a/include/sound/tlv320aic3x.h
+++ b/include/sound/tlv320aic3x.h
@@ -54,7 +54,7 @@ enum aic3x_micbias_voltage {
 };
 
 struct aic3x_setup_data {
-	unsigned int gpio_func[2];
+	unsigned int gpio_func[5];
 };
 
 struct aic3x_pdata {
-- 
2.7.4

